state-chain/pallets/cf-lending-pools/src/general_lending.rs

```diff
@@ -0,0 +1,1635 @@
+
+			let max_slippage = if is_hard {
+				config.hard_liquidation_max_oracle_slippage
+			} else {
+				config.soft_liquidation_max_oracle_slippage
+			};
+
+			let swap_request_id = initiate_swap::<T>(
+				from_asset,
+				collateral_amount,
+				to_asset,
+				LendingSwapType::Liquidation { borrower_id: borrower_id.clone(), loan_id },
+				max_slippage,
+			);
+
+			swaps_for_event.entry(loan_id).or_default().push(swap_request_id);
+
+			liquidation_swaps
+				.insert(swap_request_id, LiquidationSwap { loan_id, from_asset, to_asset });
+		}
+
+		Pallet::<T>::deposit_event(Event::LiquidationInitiated {
+			borrower_id: borrower_id.clone(),
+			swaps: swaps_for_event,
+			is_hard,
+		});
+
+		self.liquidation_status = LiquidationStatus::Liquidating { liquidation_swaps, is_hard };
+	}
+
+	/// Return collateral from aborted liquidation swaps (to be called when no further liquidation
+	/// is required).
+	fn return_collateral(&mut self, liquidation_swaps_collected: Vec<AssetCollateralForLoan>) {
+		for AssetCollateralForLoan { collateral_asset, collateral_amount, .. } in
+			liquidation_swaps_collected
+		{
+			self.collateral
+				.entry(collateral_asset)
+				.or_default()
+				.saturating_accrue(collateral_amount);
+		}
+
+		self.liquidation_status = LiquidationStatus::NoLiquidation;
+	}
+
+	fn settle_loan(&mut self, loan_id: LoanId, via_liquidation: bool) {
+		if let Some(loan) = self.loans.remove(&loan_id) {
+			Pallet::<T>::deposit_event(Event::LoanSettled {
+				loan_id,
+				total_fees: loan.fees_paid,
+				via_liquidation,
+			});
+		}
+	}
+
+	/// Repays (fully or partially) the loan with `provided_amount` (that was either debited from
+	/// the account or received during liquidation). Returns any unused amount. If the loan does not
+	/// exist, returns `None`.
+	#[transactional]
+	fn repay_principal(
+		&mut self,
+		loan_id: LoanId,
+		provided_asset: Asset,
+		provided_amount: AssetAmount,
+		should_charge_liquidation_fee: bool,
+	) -> Result<LoanRepaymentOutcome, DispatchError> {
+		let config = LendingConfig::<T>::get();
+
+		let Some(loan) = self.loans.get_mut(&loan_id) else {
+			// In rare cases it may be possible for the loan to no longer exist if
+			// e.g. the principal was fully covered by a prior liquidation swap.
+			fail!(Error::<T>::LoanNotFound);
+		};
+
+		if loan.asset != provided_asset {
+			log_or_panic!(
+				"Unexpected asset {} provided to repay loan {loan_id}, expected {}",
+				provided_asset,
+				loan.asset
+			);
+			fail!(Error::<T>::InternalInvariantViolation);
+		}
+
+		let (provided_amount_after_fees, liquidation_fee) = if should_charge_liquidation_fee {
+			let liquidation_fee =
+				config.get_config_for_asset(loan.asset).liquidation_fee * provided_amount;
+			let after_fees = provided_amount.saturating_sub(liquidation_fee);
+
+			// NOTE: this may not be equal to the provided fee if the provided amount isn't
+			// sufficient
+			let actual_fee = provided_amount.saturating_sub(after_fees);
```

Comment:
I don't understand this comment - since config.liquidation_fee is a Permill value, the calculated absolute liquidation fee is always <= provided_amount, so `actual_fee` is always the same as `liquidation_fee` here, or not?

Reply:
Can't remember what this was about... removed `actual_fee` and the comment.